src/intervals/intervalstream.jl

```diff
@@ -0,0 +1,180 @@
+                                             b::IntervalStreamOrArray{T},
+                                             seqname_isless::Function)
+        return new(a, b, seqname_isless, StreamBuffer{Interval{S}}())
+    end
+end
+
+
+immutable IntervalStreamIntersectIteratorState{U, V}
+    a_buffer_pos::Int
+    a_state::U
+    b_state::V
+end
+
+
+@doc """
+Intersect two `IntervalStreams` returning an iterator over all pairs of
+intersecting intervals.
+""" ->
+function Base.intersect{S, T}(a::IntervalStreamOrArray{S},
+                              b::IntervalStreamOrArray{T},
+                              seqname_isless::Function=isless)
+    return IntervalStreamIntersectIterator{S, T}(a, b, seqname_isless)
+end
+
+
+function first_intersection{S, T, U, V}(a::IntervalStreamOrArray{S},
+                                        b::IntervalStreamOrArray{T},
+                                        a_state::U,
+                                        b_state::V,
+                                        a_interval::Interval{S},
+                                        b_interval::Interval{T},
+                                        seqname_isless::Function)
+
+    while !done(a, a_state) || !done(b, b_state)
+        if precedes(a_interval, b_interval) && !done(a, a_state)
+            a_interval_next, a_state = next(a, a_state)
+            if isless(a_interval_next, a_interval, seqname_isless)
+                error("Interval streams must be sorted to perform intersection.")
+            end
+            a_interval = a_interval_next
+        elseif precedes(b_interval, a_interval) && !done(b, b_state)
+            b_interval_next, b_state = next(b, b_state)
+            if isless(b_interval_next, b_interval, seqname_isless)
+                error("Interval streams must be sorted to perform intersection.")
+            end
+            b_interval = b_interval_next
+        else # intersection
+            break
+        end
+    end
+
+    return a_interval, b_interval, a_state, b_state
+end
+
+
+function Base.start{S, T}(it::IntervalStreamIntersectIterator{S, T})
+    a_state = start(it.a)
+    b_state = start(it.b)
+
+    if done(it.a, a_state) || done(it.b, b_state)
+        return IntervalStreamIntersectIteratorState(a_state, b_state)
+    end
+
+    a_interval, a_state = next(it.a, a_state)
+    b_interval, b_state = next(it.b, b_state)
+    a_interval, b_interval, a_state, b_state = first_intersection(
+        it.a, it.b, a_state, b_state, a_interval, b_interval, it.seqname_isless)
+
+    if isoverlapping(a_interval, b_interval)
+        push!(it.a_buffer, a_interval)
+        it.b_interval = b_interval
+    end
+
+    return IntervalStreamIntersectIteratorState(1, a_state, b_state)
+end
+
+
+# Do we need a state if this is going to be stateful. State can maybe just
+# be the current interval in b
+function Base.next{S, T, U, V}(it::IntervalStreamIntersectIterator{S, T},
```

Comment:
This generic function is imported explicit on this module, therefore we don't need `Base.` here.